### ISSUE TITLE
Scan.devices.replace crash when running macOS tests

### ISSUE THREAD

**Issue by Cyberbeni** [lyndsey-ferguson/fastlane-plugin-test_center#291]
### Question Checklist

- [x] Updated `fastlane-plugin-test_center` to the latest version
- [x] I read the [README.md](https://github.com/lyndsey-ferguson/fastlane-plugin-test_center/blob/master/README.md)
- [x] I reviewed the [example(s)](https://github.com/lyndsey-ferguson/fastlane-plugin-test_center/blob/master/README.md) for the action(s) I am using
- [x] I searched for [existing GitHub issues](https://github.com/lyndsey-ferguson/fastlane-plugin-test_center/issues)

> If you love this fastlane plugin, consider sponsoring it or asking your company to sponsor it. I would really appreciate any
> gesture: https://github.com/sponsors/lyndsey-ferguson. 😍

#### Question Description

We are using our own fork of fastlane and test_center (that we occasionally update by merging the latest releases) for running test that can run tests without specifying a project, only derived data (but the changes break other stuff)
We got a crash at this line saying that `undefined method 'replace' for nil:NilClass`
https://github.com/lyndsey-ferguson/fastlane-plugin-test_center/blob/33a267290db7ac4371f8b939408698d4fda5557b/lib/fastlane/plugin/test_center/helper/multi_scan_manager/retrying_scan.rb#L53
I don't use ruby much but as far as I know `Scan.devices = @options[:scan_devices_override]` would have the same result minus the crashing when Scan.devices is nil at this point.
If my assumption is correct, it might be better to have a linter rule against using replace to make the project less error prone.

**Comment by lyndsey-ferguson** [comment]
You're assumption is correct, `Scan.devices` is nil at this point. Which it shouldn't be as the system _does_ think it should replace this at the moment.

I'd be interested in seeing the console logs (with the fastlane command run with `--verbose`), but I'm not sure how much customization you've done to your project.

**Comment by Cyberbeni** [comment]
This part is inside an `unless config[:xctestrun]` block (our macOS tests use the project file instead of xctestrun as they checkout the git workspace because they need to rebuild the test runner to use the correct version of XCTest): https://github.com/fastlane/fastlane/blob/48151291f2c4949c3b1b9919ba2cc81a7cc33293/scan/lib/scan/detect_values.rb#L18-L40
With an added comment: `# TODO: detect_simulator could work without a project`
There is not much customization, we just basically specified the same input we used with xcodebuild (when migrating to multi_scan against flaky ui tests), and then tried to fix the errors as we got them. These are the parameters we use:
```
    multi_scan(
      scheme: "SchemeName", # needed for scan file/dir names, no scheme is actually used as there is no project file
      output_directory: options[:output_directory],
      result_bundle: true,
      test_without_building: true,
      xctestrun: options[:xctestrun],
      destination: options[:destination],
      try_count: 3,
      quit_simulators: false,
      parallel_testrun_count: options[:parallel_count].to_i,
      batch_count: 18,
      fail_build: false,
      xcargs: '-maximum-test-execution-time-allowance 600 -test-timeouts-enabled YES',
      output_types: 'junit',
      output_files: 'result.xml',
      buildlog_path: "#{options[:output_directory]}",
      xcodebuild_log_file_name: "xcodetest.log",
      append_to_xcodebuild_log: true,
      skip_testing: []
    )
```

**Comment by lyndsey-ferguson** [comment]
@Cyberbeni, the key is what you wrote above "...our macOS tests..." indicates that I had forgotten about simulator-less tests in `multi_scan`. I've modified the code in a branch that I would appreciate your tests on.  Can you modify your `Pluginfile` per my instructions below, run `bundle install`, and then run your fastlane again (with the `--verbose` flag)?

Pluginfile:
```
gem 'fastlane-plugin-test_center', :git => "https://github.com/lyndsey-ferguson/fastlane-plugin-test_center.git", :branch => "issue-291-macos-crashing"
```

If there are still problems, please let me know and attach the console output as a text file to this issue (makes it easier for me to review).  If it works, please let me know.

**Comment by Cyberbeni** [comment]
Our macOS tests didn't have any problems after merging 3.14.4 to our fork and fixing the line linked in the original comment. (Only tried running iOS tests before the fix)

**Comment by lyndsey-ferguson** [comment]
Sure, `Scan.devices = @options[:scan_devices_override]` will work, but that should not even be done in the macOS case. Can you test my change that I mentioned above? 🙏

**Comment by lyndsey-ferguson** [comment]
Fixed in [v3.14.5](https://github.com/lyndsey-ferguson/fastlane-plugin-test_center/releases/tag/v3.14.5)

**Comment by Cyberbeni** [comment]
This (or some other change) breaks the tests on macOS:

```
bundler: failed to load command: fastlane (/Users/benedek.kozma/dev/TestApp/.gems/ruby/2.6.0/bin/fastlane)
FastlaneCore::Interface::FastlaneShellError: [!] Shell command exited with exit status 2 instead of 0.
  /Users/benedek.kozma/dev/TestApp/.gems/ruby/2.6.0/gems/fastlane-2.163.0/fastlane_core/lib/fastlane_core/ui/interface.rb:153:in `shell_error!'
  /Users/benedek.kozma/dev/TestApp/.gems/ruby/2.6.0/gems/fastlane-2.163.0/fastlane_core/lib/fastlane_core/ui/ui.rb:17:in `method_missing'
  /Users/benedek.kozma/dev/TestApp/.gems/ruby/2.6.0/gems/fastlane-2.163.0/fastlane/lib/fastlane/helper/sh_helper.rb:80:in `sh_control_output'
  /Users/benedek.kozma/dev/TestApp/.gems/ruby/2.6.0/gems/fastlane-2.163.0/fastlane/lib/fastlane/helper/sh_helper.rb:12:in `sh'
  /Users/benedek.kozma/dev/TestApp/.gems/ruby/2.6.0/gems/fastlane-plugin-test_center-3.14.5/lib/fastlane/plugin/test_center/helper/multi_scan_manager/runner.rb:127:in `block in setup_run_tests_for_each_device'
  /Users/benedek.kozma/dev/TestApp/.gems/ruby/2.6.0/gems/fastlane-plugin-test_center-3.14.5/lib/fastlane/plugin/test_center/helper/multi_scan_manager/runner.rb:118:in `each'
  /Users/benedek.kozma/dev/TestApp/.gems/ruby/2.6.0/gems/fastlane-plugin-test_center-3.14.5/lib/fastlane/plugin/test_center/helper/multi_scan_manager/runner.rb:118:in `each_with_index'
  /Users/benedek.kozma/dev/TestApp/.gems/ruby/2.6.0/gems/fastlane-plugin-test_center-3.14.5/lib/fastlane/plugin/test_center/helper/multi_scan_manager/runner.rb:118:in `setup_run_tests_for_each_device'
  /Users/benedek.kozma/dev/TestApp/.gems/ruby/2.6.0/gems/fastlane-plugin-test_center-3.14.5/lib/fastlane/plugin/test_center/helper/multi_scan_manager/runner.rb:105:in `run'
  /Users/benedek.kozma/dev/TestApp/.gems/ruby/2.6.0/gems/fastlane-plugin-test_center-3.14.5/lib/fastlane/plugin/test_center/actions/multi_scan.rb:43:in `run'
  /Users/benedek.kozma/dev/TestApp/.gems/ruby/2.6.0/gems/fastlane-2.163.0/fastlane/lib/fastlane/runner.rb:263:in `block (2 levels) in execute_action'
  /Users/benedek.kozma/dev/TestApp/.gems/ruby/2.6.0/gems/fastlane-2.163.0/fastlane/lib/fastlane/actions/actions_helper.rb:69:in `execute_action'
  /Users/benedek.kozma/dev/TestApp/.gems/ruby/2.6.0/gems/fastlane-2.163.0/fastlane/lib/fastlane/runner.rb:255:in `block in execute_action'
  /Users/benedek.kozma/dev/TestApp/.gems/ruby/2.6.0/gems/fastlane-2.163.0/fastlane/lib/fastlane/runner.rb:229:in `chdir'
  /Users/benedek.kozma/dev/TestApp/.gems/ruby/2.6.0/gems/fastlane-2.163.0/fastlane/lib/fastlane/runner.rb:229:in `execute_action'
  /Users/benedek.kozma/dev/TestApp/.gems/ruby/2.6.0/gems/fastlane-2.163.0/fastlane/lib/fastlane/runner.rb:157:in `trigger_action_by_name'
  /Users/benedek.kozma/dev/TestApp/.gems/ruby/2.6.0/gems/fastlane-2.163.0/fastlane/lib/fastlane/fast_file.rb:159:in `method_missing'
  Fastfile:21:in `block (2 levels) in parsing_binding'
  /Users/benedek.kozma/dev/TestApp/.gems/ruby/2.6.0/gems/fastlane-2.163.0/fastlane/lib/fastlane/lane.rb:33:in `call'
  /Users/benedek.kozma/dev/TestApp/.gems/ruby/2.6.0/gems/fastlane-2.163.0/fastlane/lib/fastlane/runner.rb:49:in `block in execute'
  /Users/benedek.kozma/dev/TestApp/.gems/ruby/2.6.0/gems/fastlane-2.163.0/fastlane/lib/fastlane/runner.rb:45:in `chdir'
  /Users/benedek.kozma/dev/TestApp/.gems/ruby/2.6.0/gems/fastlane-2.163.0/fastlane/lib/fastlane/runner.rb:45:in `execute'
  /Users/benedek.kozma/dev/TestApp/.gems/ruby/2.6.0/gems/fastlane-2.163.0/fastlane/lib/fastlane/lane_manager.rb:47:in `cruise_lane'
  /Users/benedek.kozma/dev/TestApp/.gems/ruby/2.6.0/gems/fastlane-2.163.0/fastlane/lib/fastlane/command_line_handler.rb:36:in `handle'
  /Users/benedek.kozma/dev/TestApp/.gems/ruby/2.6.0/gems/fastlane-2.163.0/fastlane/lib/fastlane/commands_generator.rb:108:in `block (2 levels) in run'
  /Users/benedek.kozma/dev/TestApp/.gems/ruby/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/command.rb:178:in `call'
  /Users/benedek.kozma/dev/TestApp/.gems/ruby/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/command.rb:153:in `run'
  /Users/benedek.kozma/dev/TestApp/.gems/ruby/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/runner.rb:476:in `run_active_command'
  /Users/benedek.kozma/dev/TestApp/.gems/ruby/2.6.0/gems/fastlane-2.163.0/fastlane_core/lib/fastlane_core/ui/fastlane_runner.rb:76:in `run!'
  /Users/benedek.kozma/dev/TestApp/.gems/ruby/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/delegates.rb:15:in `run!'
  /Users/benedek.kozma/dev/TestApp/.gems/ruby/2.6.0/gems/fastlane-2.163.0/fastlane/lib/fastlane/commands_generator.rb:352:in `run'
  /Users/benedek.kozma/dev/TestApp/.gems/ruby/2.6.0/gems/fastlane-2.163.0/fastlane/lib/fastlane/commands_generator.rb:41:in `start'
  /Users/benedek.kozma/dev/TestApp/.gems/ruby/2.6.0/gems/fastlane-2.163.0/fastlane/lib/fastlane/cli_tools_distributor.rb:119:in `take_off'
  /Users/benedek.kozma/dev/TestApp/.gems/ruby/2.6.0/gems/fastlane-2.163.0/bin/fastlane:23:in `<top (required)>'
  /Users/benedek.kozma/dev/TestApp/.gems/ruby/2.6.0/bin/fastlane:23:in `load'
  /Users/benedek.kozma/dev/TestApp/.gems/ruby/2.6.0/bin/fastlane:23:in `<top (required)>'
```

**Comment by lyndsey-ferguson** [comment]
Sorry @Cyberbeni, I broke this with the recent merge.

Can you modify your `Pluginfile` per my instructions below, run `bundle install`, and then run your fastlane again (with the `--verbose` flag)?

Pluginfile:
```
gem 'fastlane-plugin-test_center', :git => "https://github.com/lyndsey-ferguson/fastlane-plugin-test_center.git", :branch => "issue-291-follow-up-new-macos-tests-crash"
```

If there are still problems, please let me know and attach the console output as a text file to this issue (makes it easier for me to review).  If it works, please let me know.

**Comment by Cyberbeni** [comment]
It fixes the issue.

**Comment by lyndsey-ferguson** [comment]
Released in [v3.14.6](https://github.com/lyndsey-ferguson/fastlane-plugin-test_center/releases/tag/v3.14.6)
